### PR TITLE
Game play/achievements

### PR DESCRIPTION
**Pull request by BhavyamKrGoyal**
with scriptable objects for achievements and unlocking the same

### CODE REVIEW

File: Assets/scripts/ScriptableObjects/Achievement.cs [hunk 1]
```diff
@@ -12,8 +12,9 @@ public class Achievement : ScriptableObject
         public bool sessionBased;
 
 
-        public void UpdateAchievement(int progress, ControllerPlayer player)
+        public string UpdateAchievement(int progress, ControllerPlayer player)
         {
+            string result = null;
             //Debug.Log("Updating achievement: " + achievementDisplayName);
             int currentLevel = PlayerPrefs.GetInt(achievementDisplayName + player.GetPlayerNumber() + "level", 0);
```
Comment (SwapnilOutscal) [inline Assets/scripts/ScriptableObjects/Achievement.cs:23]: Use a separate class for Saving and loading the data.

File: Assets/scripts/Achievements/AchievementLevel.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿using UnityEngine;
+using UnityEditor;
+[System.Serializable]
```
Comment (mayankgrover) [inline Assets/scripts/Achievements/AchievementLevel.cs]: namespace

File: Assets/scripts/Achievements/ServiceAchievements.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+           
+        }
+        public void AddListener(ControllerPlayer player)
+        {
+            PlayerData playerData = player.playerData;
+            player.OnBulletShot += AchievementUpdate;
+            player.OnUIUpdate += AchievementUpdate;
+            player.OnEnemyKilled += AchievementUpdate;
+            //Debug.Log("All listeners added for achievements");
+            if (player.GetGameStarted())
```
Comment (mayankgrover) [inline Assets/scripts/Achievements/ServiceAchievements.cs:51]: i don't understand why achievement types is dependent on player.getgamesstarted ?

File: Assets/scripts/Enums.cs [hunk 4]
```diff
@@ -11,4 +11,20 @@ public enum Controls
     WASD = 1,
     IJKL = 2
 }
+public enum PlayerNumber
```
Comment (mayankgrover) [inline Assets/scripts/Enums.cs:14]: create separate files for each enum in the correct namespaces.

File: Assets/scripts/Enemy/View/ViewEnemy.cs [hunk 5]
```diff
@@ -30,7 +30,7 @@ public Vector3 GetPosition()
     }
     public void TakeDamage(int damage,IBasePlayerController player)
     {
-        controller.BulletHit(damage,player);
+        controller.BulletHit(damage,(ControllerPlayer)player);
```
Comment (mayankgrover) [inline Assets/scripts/Enemy/View/ViewEnemy.cs:33]: point of interfaces is that you don't need to know about controllerplayer implementation in view enemy but now you have this hard coupled. work with interfaces

File: Assets/scripts/Enums.cs [hunk 6]
```diff
@@ -11,4 +11,20 @@ public enum Controls
...
+{
+    Player1 = 1,
+    Player2 = 2,
+    Player3 = 3,
+    Player4 = 4
+}
+public enum AchievementTypes
+{
+    BulletsShot=1,
```
Comment (mayankgrover) [inline Assets/scripts/Enums.cs:29]: if its always 1,2,3,.. then the numbers are not needed.

File: Assets/scripts/ScriptableObjects/Achievement.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        public bool sessionBased;
+
+
+        public string UpdateAchievement(int progress, ControllerPlayer player)
+        {
+            string result = null;
+            //Debug.Log("Updating achievement: " + achievementDisplayName);
+            int currentLevel = PlayerPrefs.GetInt(achievementDisplayName + player.GetPlayerNumber() + "level", 0);
+            
+            if (!sessionBased)
```
Comment (mayankgrover) [inline Assets/scripts/ScriptableObjects/Achievement.cs]: simplify the whole logic like we discussed

File: Assets/scripts/UI/View/ViewStartUI.cs [hunk 8]
```diff
@@ -8,15 +9,22 @@ public class ViewStartUI : MonoBehaviour
 {
     Text score;
     Text health;
+    Text achievement;
     public void Start()
     {
+        if (ServiceAchievements.Instance != null)
```
Comment (mayankgrover) [inline Assets/scripts/UI/View/ViewStartUI.cs:17]: was this throwing a NPE ?

File: Assets/scripts/UI/View/ViewStartUI.cs [hunk 9]
```diff
@@ -8,15 +9,22 @@ public class ViewStartUI : MonoBehaviour
...
+        {
+            ServiceAchievements.Instance.OnAchievementUnlocked += AchievementUnlocked;
+        }
         score = GameObject.FindGameObjectWithTag("Score").GetComponent<Text>();
         health = GameObject.FindGameObjectWithTag("Health").GetComponent<Text>();
+        achievement= GameObject.FindGameObjectWithTag("AchievementUI").GetComponent<Text>();
         DestroyUI();
     }
 
     public void DestroyUI()
```
Comment (mayankgrover) [inline Assets/scripts/UI/View/ViewStartUI.cs:29]: ui is disabled but the function is called destroy

File: Assets/scripts/structs.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using UnityEngine;
+using UnityEditor;
+
+public struct PlayerData
```
Comment (mayankgrover) [inline Assets/scripts/structs.cs:4]: namespace :)

**Review by mayankgrover — CHANGES_REQUESTED** [review]

File: Assets/scripts/Achievements/ServiceAchievements.cs (hunk 3 above)
Comment (BhavyamKrGoyal) [reply Assets/scripts/Achievements/ServiceAchievements.cs:51]: get game started is to check that player has started the game not respawned

File: Assets/scripts/UI/View/ViewStartUI.cs (hunk 8 above)
Comment (BhavyamKrGoyal) [reply Assets/scripts/UI/View/ViewStartUI.cs:17]: I did it for a case if I disable achievements.

File: Assets/scripts/ScriptableObjects/Achievement.cs (hunk 1 above)
Comment (BhavyamKrGoyal) [reply Assets/scripts/ScriptableObjects/Achievement.cs:23]: done in save system

File: Assets/scripts/Achievements/AchievementLevel.cs (hunk 2 above)
Comment (BhavyamKrGoyal) [reply Assets/scripts/Achievements/AchievementLevel.cs]: done

File: Assets/scripts/structs.cs (hunk 10 above)
Comment (BhavyamKrGoyal) [reply Assets/scripts/structs.cs:4]: done

**Pull request merged**